game.py

```diff
@@ -198,146 +263,88 @@ def testPiecesPurchasable(self):
-    #     if not self.canBuyPiece(player_num, piecetype): 
-    #         print("You don't have enough resources to buy this piece")
-    #         return
-    #     resources_needed = defaultdict(str)
-    #     #Put values in resources needed based on the piece type
-    #     if piecetype == "Settlement":
-    #         resources_needed['Brick'] = 1
-    #         resources_needed['Wood'] = 1
-    #         resources_needed['Wool'] = 1
-    #         resources_needed['Wheat'] = 1
-    #     elif piecetype == "City":
-    #         resources_needed['Ore'] = 3
-    #         resources_needed['Wheat'] = 2
-    #     elif piecetype == "Road":
-    #         resources_needed['Brick'] = 1
-    #         resources_needed['Wood'] = 1
-    
-    #     #Get resources the player has
-    #     cur_resources = self.players[player_num].resources
-
-    #     #Update players resources
-    #     for key in resources_needed:
-    #         cur_resources[key] -= resources_needed[key]
-
-    #     #Give piece to player
-    #     self.players[player_num].pieces[piecetype] += 1
-
-    ###Need better understanding of board architecture to implement these
-    # def canPlacePiece()
-    # def getAvailableLocations()        
-    # def placePiece():
-
-    ################################################################
-    ######################   Get Actions   ########################
-    ################################################################
-    '''
-    Returns a list of all possible actions for a given game.
-    Pretty much all of the information needed should be stored in the game class.
-    Makes use of helpers for different types of actions.
-    Possible action types:
-        -buyPiece: (placing a piece is incorporated as part of buying a piece)
-            -Road
-            -Settlement
-            -City
-        -buyDevCard
-        -playDevCard
-            -Knight
-            -Victory Point
-            -etc
-        -end turn
-        -trade (to be implemented later)
-
-
-        Return format:
-        [(piece, location)]
-    '''
 
     '''
-    Current thinking: 
-        -Create functions that give a list of possible [(piece, location)] assignments for a given gamestate
-        -Functions also return number of these assignments you can pick
+    This section handles getting possible locations for different types of pieces and given a 
+    certian player. 
     '''
-
+    
     #Get valid road locations
     def getRoadLocations(self, player):
         possible_locations = []
-        
+        pass
         #Loop over all edges, check if edge is empty and one of the neighbors is a valid city/road
             #Add valid edges to the list
-        
         return possible_locations
     
-    #Get valid bu
+    '''Missing test for settlement being at end of road'''
+    #Helper to test if node is valid for a settlment. 
+    def isValidSettlement(node):
+        if node.isOccupied or node.neighbors: 
```

Comment:
does the check for node.neighbors work like this or do we need to check each one to see if it has something in it? the node will always have neighbors I think we need to check if they contain anything

Reply:
yeah you're right